### PR TITLE
Remove obsolete-marked APIs for .NET v2 SDK

### PR DESCRIPTION
**Pull request by brycewang-microsoft**

### CODE REVIEW

File: iothub/service/src/Common/Extensions/CommonExtensions.cs [hunk 1]
```diff
@@ -30,79 +30,14 @@ public static class CommonExtensionMethods
         private const char ValuePairDelimiter = ';';
         private const char ValuePairSeparator = '=';
 
-        /// <summary>
```
Comment (brycewang-microsoft) [inline iothub/service/src/Common/Extensions/CommonExtensions.cs:33]: Methods gets removed as they don't have any references, here and below.

File: iothub/service/src/Common/Extensions/CommonExtensions.cs [hunk 2]
```diff
@@ -30,79 +30,14 @@ public static class CommonExtensionMethods
...
-        /// Appends the specified character, if not already there
-        /// </summary>
-        /// <param name="value">The string value to update</param>
-        /// <param name="suffix">The desired suffix</param>
-        /// <returns>The fixed up string</returns>
-        public static string EnsureEndsWith(this string value, char suffix)
```
Comment (brycewang-microsoft) [inline iothub/service/src/Common/Extensions/CommonExtensions.cs:39]: Methods gets removed as they don't have any references, here and below.

File: iothub/service/src/Common/Extensions/CommonExtensions.cs [hunk 3]
```diff
@@ -30,79 +30,14 @@ public static class CommonExtensionMethods
...
-        }
-
         /// <summary>
         /// Takes a string representation of key/value pairs and produces a dictionary
         /// </summary>
         /// <param name="valuePairString">The string containing key/value pairs</param>
         /// <param name="kvpDelimiter">The delimeter between key/value pairs</param>
         /// <param name="kvpSeparator">The character separating each key and value</param>
         /// <returns>A dictionary of the key/value pairs</returns>
-        public static IDictionary<string, string> ToDictionary(this string valuePairString, char kvpDelimiter, char kvpSeparator)
```
Comment (brycewang-microsoft) [inline iothub/service/src/Common/Extensions/CommonExtensions.cs:105]: This class is not supported for external use.

File: iothub/service/src/Common/Extensions/DictionaryExtensions.cs [hunk 4]
```diff
@@ -1,106 +0,0 @@
-// Copyright (c) Microsoft. All rights reserved.
```
Comment (brycewang-microsoft) [inline iothub/service/src/Common/Extensions/DictionaryExtensions.cs:1]: None of methods within this class has any references.

File: iothub/service/src/Common/Extensions/ExceptionExtensions.cs [hunk 5]
```diff
@@ -23,53 +23,17 @@ public static class ExceptionExtensions
         /// </summary>
         /// <param name="exception">The exception to test.</param>
         /// <returns>True if the exception is considered fatal; otherwise, false.</returns>
-        public static bool IsFatal(this Exception exception)
```
Comment (brycewang-microsoft) [inline iothub/service/src/Common/Extensions/ExceptionExtensions.cs:26]: This method is not recommended for external use.

File: iothub/service/src/Common/Extensions/ExceptionExtensions.cs [hunk 6]
```diff
@@ -23,53 +23,17 @@ public static class ExceptionExtensions
...
         {
             return Fx.IsFatal(exception);
         }
 
-        /// <summary>
-        /// Unwinds the inner exceptions from a given exception.
-        /// </summary>
-        /// <param name="exception">The exception to unwind.</param>
-        /// <returns>The list of inner exceptions within the specified exception.</returns>
-        public static IEnumerable<Exception> Unwind(this Exception exception)
```
Comment (brycewang-microsoft) [inline iothub/service/src/Common/Extensions/ExceptionExtensions.cs:36]: Methods without any references.

File: iothub/service/src/Common/Extensions/OtherExtensions.cs [hunk 7]
```diff
@@ -14,27 +14,13 @@ namespace Microsoft.Azure.Devices.Common.Extensions
...
     public static class OtherExtensions
     {
-        /// <summary>
-        /// Gets the value associated with specified key.
-        /// </summary>
-        /// <typeparam name="T">The type of the value parameter.</typeparam>
-        /// <param name="dictionary">The dictionary containing the specified key.</param>
-        /// <param name="key">The key whose value to get.</param>
-        /// <returns>The value associated with the specified key, if the key is found; otherwise, the default value for the type of the value parameter.</returns>
-        public static T GetValueOrDefault<T>(this IDictionary<string, object> dictionary, string key)
```
Comment (brycewang-microsoft) [inline iothub/service/src/Common/Extensions/OtherExtensions.cs:24]: Methods without any references

File: iothub/service/src/RegistryManager.cs [hunk 8]
```diff
@@ -406,46 +406,6 @@ public virtual Task<BulkRegistryOperationResult> AddDeviceWithTwinAsync(Device d
             }
         }
 
-        /// <summary>
-        /// Register a list of new devices with the system
-        /// </summary>
-        /// <param name="devices">The Device objects being registered.</param>
-        /// <returns>Returns a string array of error messages.</returns>
-        [Obsolete("Use AddDevices2Async")]
-        public virtual Task<string[]> AddDevicesAsync(IEnumerable<Device> devices)
```
Comment (brycewang-microsoft) [inline iothub/service/src/RegistryManager.cs:415]: **Question**: in this class, we have multiple methods(like `AddDevicesAsync()` ) which have replacements (`AddDevices2Async()`). With the deprecated ones removed, do we wanna rename the replacements to the original ones (`AddDevices2Async()` -> `AddDevicesAsync()`)?
@abhipsaMisra @timtay-microsoft

Comment (abhipsaMisra) [reply iothub/service/src/RegistryManager.cs:415]: That sounds like a good suggestion.

File: e2e/test/iothub/method/MethodE2ETests.cs [hunk 9]
```diff
@@ -477,12 +477,12 @@ public static async Task<Task> SetDeviceReceiveMethodDefaultHandlerAsync(DeviceC
             return methodCallReceived.Task;
         }
 
-        private static Task<Task> SetDeviceReceiveMethodObsoleteHandler(DeviceClient deviceClient, string methodName, MsTestLogger logger)
+        private static async Task<Task> SetDeviceReceiveMethodObsoleteHandler(DeviceClient deviceClient, string methodName, MsTestLogger logger)
```
Comment (abhipsaMisra) [inline e2e/test/iothub/method/MethodE2ETests.cs:480]: nit: we could rename the test method here to remove the "obsolete" keyword and add the "async" suffix.

Comment (abhipsaMisra) [reply e2e/test/iothub/method/MethodE2ETests.cs:480]: On that note, we should be having an existing test method for testing this device client method, so you might be able to simply delete this one.

File: iothub/device/src/DeviceClient.cs [hunk 10]
```diff
@@ -571,27 +561,6 @@ public void SetRetryPolicy(IRetryPolicy retryPolicy)
...
             InternalClient.SendEventBatchAsync(messages, cancellationToken);
 
-        /// <summary>
-        /// Uploads a stream to a block blob in a storage account associated with the IoTHub for that device.
-        /// If the blob already exists, it will be overwritten.
-        /// </summary>
-        /// <param name="blobName">The name of the blob to upload.</param>
-        /// <param name="source">A stream with blob contents. Should be disposed after upload completes.</param>
-        [Obsolete("This API has been split into three APIs: GetFileUploadSasUri, uploading to blob directly using the Azure Storage SDK, and CompleteFileUploadAsync")]
-        public Task UploadToBlobAsync(string blobName, Stream source) => InternalClient.UploadToBlobAsync(blobName, source);
```
Comment (abhipsaMisra) [inline iothub/device/src/DeviceClient.cs:581]: Another improvement would be to consolidate the cancellation token and non cancellation token variants of the methods into a single method with an optional cancellation token. You could take it up in a different PR though.

File: iothub/device/tests/Amqp/AmqpTransportHandlerTests.cs [hunk 11]
```diff
@@ -95,34 +95,6 @@ public async Task AmqpTransport_Select_CorrectReceiverLink_ForModuleTwin()
...
-        [TestMethod]
-        [Obsolete]
-        public void AmqpTransportHandler_RejectAmqpSettingsChange()
-        {
-            // Added [Obsolete] attribute to this method to suppress CS0618 message for ConnectionIdleTimeout
-            var amqpTransportHandler1 = new AmqpTransportHandler(new PipelineContext(), new IotHubConnectionString(IotHubConnectionStringBuilder.Create(TestConnectionString)), new AmqpTransportSettings(TransportType.Amqp_Tcp_Only, 60, new AmqpConnectionPoolSettings()
-            {
-                Pooling = true,
-                MaxPoolSize = 10,
-                ConnectionIdleTimeout = TimeSpan.FromMinutes(1)
```
Comment (abhipsaMisra) [inline iothub/device/tests/Amqp/AmqpTransportHandlerTests.cs:107]: For some context, this test checks that amqp settings changes are rejected, which is still a valid scenario. The obsolote part of this test is the fact that ConnectionIdleTimeout is being set as one of the setting's properties. You could simply remove ConnectionIdleTimeout and remove the obsolete tag.

File: iothub/device/tests/DeviceClientTests.cs [hunk 12]
```diff
@@ -425,7 +425,7 @@ public async Task DeviceClient_OnMethodCalled_NullMethodRequest_With_SetMethodHa
 
             bool isMethodHandlerCalled = false;
 #pragma warning disable CS0618 // Type or member is obsolete
-            deviceClient.SetMethodHandler("testMethodName", (payload, context) =>
```
Comment (abhipsaMisra) [inline iothub/device/tests/DeviceClientTests.cs:428]: I'd recommend doing a sanity check to ensure we don't have an existing test testing the `deviceClient.SetMethodHandlerAsync` method. If we do, then the obsolete test can be simply removed.

File: iothub/device/tests/DeviceClientTwinApiTests.cs [hunk 13]
```diff
@@ -28,7 +28,7 @@ public async Task DeviceClientSetDesiredPropertyUpdateCallbackRegistersForPatche
 
             // act
 #pragma warning disable CS0618 // Type or member is obsolete
-            await client.SetDesiredPropertyUpdateCallback(myCallback, context).ConfigureAwait(false);
```
Comment (abhipsaMisra) [inline iothub/device/tests/DeviceClientTwinApiTests.cs:31]: same sanity check here and other test classes as well

**Review by abhipsaMisra — APPROVED** [review]

File: iothub/service/src/AmqpFeedbackReceiver.cs [hunk 14]
```diff
@@ -61,19 +61,6 @@ public Task CloseAsync()
             }
         }
 
-        [Obsolete("Use ReceiveAsync(CancellationToken cancellationToken).")]
-        public override Task<FeedbackBatch> ReceiveAsync()
```
[the remark this answers is not in the document]
Comment (brycewang-microsoft) [reply iothub/service/src/AmqpFeedbackReceiver.cs:65]: > Q - is the non-cancellation token ReceiveAsync API marked as obsolete in the DeviceClient class?

No. We still keep `ReceiveAsync` API in "DeviceClient" class as following:
* `ReceiveAsync()` - [ref](https://github.com/Azure/azure-iot-sdk-csharp/blob/1725c543d1d0d3f7c511420047b22cafab706beb/iothub/device/src/DeviceClient.cs#L327)
* `ReceiveAsync(TimeSpan timeout)` - [ref](https://github.com/Azure/azure-iot-sdk-csharp/blob/1725c543d1d0d3f7c511420047b22cafab706beb/iothub/device/src/DeviceClient.cs#L356)

**Comment by timtay-microsoft** [comment]
We should be able to remove the Azure Storage SDK dependency from the device client package now, FYI

**Comment by brycewang-microsoft** [comment]
> Also, I'm not sure why so many commits are showing in this PR. Can you make sure to pull the latest from previews/v2 before you do this merge?

Most of commits are from this [PR ](https://github.com/Azure/azure-iot-sdk-csharp/pull/2410). I rebased the current branch onto "previews/v2" after your PR merged in. But sure, I can pull the latest from previews/v2 again before doing the merge.

**Comment by abhipsaMisra** [comment]
> @abhipsaMisra are there any useful tools in VS that Bryce can use to find unreferenced code?

I'm not aware of native support in Visual Studio for this but there are 3rd party code analysis extensions that you could use. An easier alternative would be to create a code analysis rule set and run that on our projects. This entry from stackoverflow explains it in detail: https://stackoverflow.com/a/37579734

File: e2e/test/E2ETests.csproj [hunk 15]
```diff
@@ -24,6 +24,7 @@
     <PackageReference Include="MSTest.TestFramework" Version="2.0.0" />
     <PackageReference Include="Microsoft.CSharp" Version="4.7" />
     <PackageReference Include="System.Diagnostics.Tracing" Version="4.3.0" />
+    <PackageReference Include="WindowsAzure.Storage" Version="9.3.2" />
```
Comment (drwill-ms) [inline e2e/test/E2ETests.csproj:27]: We should use the latest SDK, not this old and deprecated one. That might be another work item and PR, but we should track it.

Look for `Azure.Storage.*` packages.

File: iothub/service/src/Common/Extensions/CommonExtensions.cs [hunk 16]
```diff
@@ -194,19 +80,9 @@ public static void AppendKeyValuePairIfNotEmpty(this StringBuilder builder, stri
         /// </summary>
         /// <param name="value">The value to check.</param>
         /// <returns>Returns true if the value is null or empty, otherwise returns false.</returns>
-        public static bool IsNullOrWhiteSpace(this string value)
+        internal static bool IsNullOrWhiteSpace(this string value)
```
Comment (drwill-ms) [inline iothub/service/src/Common/Extensions/CommonExtensions.cs:83]: Kind of a dumb helper method. We should just natively use `stirng.IsNullOrWhiteSpace` instead.

**Review by drwill-ms — APPROVED** [review]

**Pull request merged**